### PR TITLE
Fix stale parameter index map in DistributedOptimizer

### LINKED ISSUE #2777: [BUG] DistributedOptimizer issues in Mixed-Precision: Stale model_param_group_index_map and .copy_ failure during get/load parameter_state
**Describe the bug**

During mixed-precision training (BF16 & FP32, FP8 is optional), a RuntimeError (size mismatch) is triggered when saving a checkpoint via `get_parameter_state_dp_zero`. 

Image

Based on our analysis, the hardcoded parameter reordering in `_build_model_and_main_param_groups` within `DistributedOptimizer` causes the `self.model_param_group_index_map` to become out of sync with the actual `optimizer.param_groups`.

**Steps/Code to reproduce bug**

1. Before building the DDP model and optimizer，use `Float16Module` to wrap the model for BF16 training, but manually promote certain modules (both params and inputs) to FP32.
2. Train for several steps.
3. Call `get_parameter_state_dp_zero` in `DistributedOptimizer` to collect optimizer states, which triggers the size mismatch error.

**Root Cause Analysis**

1. Initial Map Construction: In `__init__`, `self.model_param_group_index_map` is first constructed via `_build_optimizer_group_ranges`. This map records the position `(group_index, group_order)` of param in param_groups

2. Hardcoded Reordering: Subsequently, `_build_model_and_main_param_groups` reorders the parameters within each group (placing native FP32 shards at the front and main parameter shards converted from FP16/BF16 at the back) and updates `optimizer.param_groups` accordingly: 

Image

3. Index Invalidation: The `model_param_group_index_map` is not updated after this reordering. Consequently, downstream functions like `_get_main_param_and_optimizer_states` retrieve the wrong Tensors using stale group_order indices, leading to shape mismatches during buffer copy operations.

**Additional question**

1. What is the design motivation behind this specific reordering (grouping by DType)?
2. What is the recommended best practice to fix this: disabling the reordering to maintain discovery order consistency, or explicitly updating the `model_param_group_index_map` after the reordering is performed?

### PR DESCRIPTION
**Pull request by ahmadki**
# What does this PR do ?

During mixed-precision training, model_param_group_index_map becomes stale after parameters are reordered by dtype. The map is initialized based on original parameter order, but _build_model_and_main_param_groups reorders parameters (FP32 first, then FP16/BF16) for checkpoint consistency. The map was never updated to reflect this reordering. This caused get_parameter_state_dp_zero to access wrong parameters when saving checkpoints, resulting in size mismatch errors.

Solution: Rebuild model_param_group_index_map after parameter reordering to keep it synchronized with optimizer.param_groups.## Contribution process

Fixes https://github.com/NVIDIA/Megatron-LM/issues/2777

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @megatron-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Review by Phlip79 — APPROVED** [review]
Thanks for the comments and tests!

File: megatron/core/optimizer/distrib_optimizer.py [hunk 1]
```diff
@@ -604,6 +606,23 @@ def __init__(
...
+            for model_param in group_range["params"]:
+                if model_param.type() == 'torch.cuda.FloatTensor':
+                    self.model_param_group_index_map[model_param] = (group_index, param_order)
+                    param_order += 1
+            # Then, add FP16/BF16 params (in the same order as they appear in group_range["params"])
+            for model_param in group_range["params"]:
+                if model_param.type() in ['torch.cuda.HalfTensor', 'torch.cuda.BFloat16Tensor']:
+                    self.model_param_group_index_map[model_param] = (group_index, param_order)
+                    param_order += 1
+
```
Comment (asolergi-nv) [inline megatron/core/optimizer/distrib_optimizer.py]: Wouldn't it be possible to directly build model_param_group_index_map in the desired order in [_build_optimizer_group_ranges](https://github.com/ahmadki/Megatron-LM/blob/bfe8a04c5424b5d6748837c6b26feb1929edc312/megatron/core/optimizer/distrib_optimizer.py#L585-L624)?

**Comment by asolergi-nv** [comment]
/ok to test [537fa1a](https://github.com/NVIDIA/Megatron-LM/pull/2872/commits/537fa1ae174073e4825cbf7df9dba3098609b695)

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -604,6 +606,23 @@ def __init__(
             self.optimizer.param_groups = [g["orig_group"] for g in self.opt_group_ranges]
             self.optimizer.load_state_dict(self.optimizer.state_dict())
 
+        # Rebuild model_param_group_index_map to reflect parameter reordering.
+        # The _build_model_and_main_param_groups method reorders parameters by dtype
+        # (FP32 first, then FP16/BF16), so we need to update the mapping to match
+        # the new positions in optimizer.param_groups.
+        for group_index, group_range in enumerate(self.opt_group_ranges):
+            param_order = 0
```
Comment (deepakn94) [inline megatron/core/optimizer/distrib_optimizer.py]: Call this `param_idx`?

File: megatron/core/optimizer/distrib_optimizer.py [hunk 3]
```diff
@@ -604,6 +606,23 @@ def __init__(
...
+            # First, add FP32 params (in the same order as they appear in group_range["params"])
+            for model_param in group_range["params"]:
+                if model_param.type() == 'torch.cuda.FloatTensor':
+                    self.model_param_group_index_map[model_param] = (group_index, param_order)
```
Comment (deepakn94) [inline megatron/core/optimizer/distrib_optimizer.py]: Does this have repercussions on DDP's `param_index_map` too?
